gh-125286: Bug fix for trace-refs logic and shared objects

nascheme opened the pull request:
Fix the TraceRefs build to handle interned strings (both mortal and immortal) that are shared between interpreters.

<!-- gh-issue-number: gh-125286 -->
* Issue: gh-125286
<!-- /gh-issue-number -->

nascheme committed:
gh-125286: Add test for single-phase init and shared objects.

nascheme committed:
gh-125286: Bug fix for trace-refs logic and shared objects.

ericsnowcurrently commented on Objects/object.c:
```diff
@@ -2545,6 +2528,25 @@ _Py_ForgetReference(PyObject *op)
+        }
+        PyInterpreterState *main_interp = _PyInterpreterState_Main();
+        if (interp != main_interp  &&
+            interp->feature_flags & Py_RTFLAGS_USE_MAIN_OBMALLOC) {
```
FYI, you can also use `_Py_IsMainInterpreter()`:

```suggestion
        if (_Py_IsMainInterpreter(interp)  &&
            interp->feature_flags & Py_RTFLAGS_USE_MAIN_OBMALLOC) {
```

ericsnowcurrently reviewed (commented):
I'm a little unclear on the change here.  My understanding is that we don't want the interned strings on legacy subinterpreters to be tracked on those interpreter's refs.  It seems to me like the change in this PR has stopped fixing that for each interpreter, which would put us back to where we were before.  What am I missing?

Other than that, I have one small suggestion.

nascheme committed:
Slightly cleaner fix, more comments.

The case with shared objects (other than interned strings) can't seem to
happen so comment that code out.  Small code cleanups.  Simplify the
_testembed.c test, don't need a separate helper module.
